### PR TITLE
Docs: add info about exceptions usage

### PR DESCRIPTION
**Pull request by tsusanka**
We did some brainstorming with @matejcik and this is a draft of some notes on how to use exceptions in Core. What do you think? We should all be on board on this.

Updates https://github.com/trezor/trezor-firmware/issues/332

### CODE REVIEW

**Comment by matejcik** [comment]
@andrewkozlik raised an important point: in case we are simulating context with DUMMY_CONTEXT, there is a legitimate use-case for catching `wire.Error` subclasses (from the POV of the code simulating the context, it's as if it got the error "from the wire" and might want to handle it.)

the language around "MUST NOT be caught" should be less strict to reflect this and possible similar situations

**Comment by tsusanka** [comment]
@matejcik I have replaced MUST with SHOULD regarding the `wire.errors`. Do you want to specify the case you are mentioning somewhat more? It seems to me we should not confuse the reader by describing something low level and a bit unusual.

**Comment by tsusanka** [comment]
Ping @matejcik any comments on this?

File: docs/core/misc/exceptions.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+From version 2.3.0 we try to follow few rules about how we use exceptions. All new
+code MUST follow these.
+
+## Usage
+
+You MAY use any exceptions in Core's logic. Exceptions from `wire.errors` SHOULD be
+the final exceptions that are thrown and SHOULD NOT be caught. Note that `wire.Error`
+is a type of exception that is intended to be sent out over the wire. It should only
+be used in contexts where that behavior is appropriate.
```
Comment (matejcik) [inline docs/core/misc/exceptions.md]: Custom exception type hierarchies SHOULD always be derived directly from `Exception`. They SHOULD NOT be derived from other built-in exceptions (such as ValueError, TypeError, etc.)

Deriving a custom exception type signals an intention to catch and handle it somewhere in the code. For this reason, custom exception types SHOULD NOT be derived from `wire.Error` and subclasses.

**Comment by matejcik** [comment]
I had to go back to our conversation to know what's missing here: I added the bit about deriving exceptions from things like ValueError, and about deriving from wire.Error.

I'm on the fence about the DUMMY_CONTEXT thing, but I lean towards agreeing with you that we do not need to complicate this document now.

**Comment by tsusanka** [comment]
Thanks. Added!

**Pull request merged**